### ISSUE TITLE
Off by one error in FspFileSystemReadDirectoryBuffer

### ISSUE THREAD

**Issue by pmaxwell-iam** [billziss-gh/winfsp#351]
FspFileSystemReadDirectoryBuffer, a function in winfsp.h, has an off-by-one error related to searches which supply a Marker where the Marker is not found in the buffer. The first DirInfo entry of the returned buffer is ignored.

I at one point believed I had replicated this with passthrough, but if I did this must have been at a point where the directory buffer had been invalidated somehow, it has been resistant to repeated replication.

The context was a folder deletion where the earlier files in the folder have already been removed. The test case had 100 files in the folder; having deleted 0-53 it then called QueryDirectory with marker "53.txt", got back a buffer containing 54-99 and continued by deleting 55 onward.

**Comment by billziss-gh** [comment]
Thanks for the report.

It would be good to have a repro so that we can be certain. However based on your description and after looking at the "DirectoryBuffer" source code I have identified this [location](https://github.com/billziss-gh/winfsp/blob/v1.8/src/dll/dirbuf.c#L378) as potentially suspect: I note that the `IndexNum` variable at the location gets incremented regardless of whether the `Marker` is found or not.

I do not currently have the mental model that I had at the time that I wrote the code to fully assess if blindly incrementing the variable is correct or not. However looking at the binary search function at [`FspFileSystemSearchDirectoryBuffer`](https://github.com/billziss-gh/winfsp/blob/v1.8/src/dll/dirbuf.c#L88-L119) it seems to always return the index of the value less than the `Marker` when the `Marker` is not found. So it seems that incrementing the `IndexNum` variable on return from `FspFileSystemSearchDirectoryBuffer` should be correct here.

There are two cases:

- If the buffer contains the entries "A", "B" and "C" and the `Marker` is "B", the `Marker` will be found and the `IndexNum` will be 1 on return from `FspFileSystemSearchDirectoryBuffer`. It will then be incremented to 2 by the code at this [location](https://github.com/billziss-gh/winfsp/blob/v1.8/src/dll/dirbuf.c#L378), which seems correct.

- If the buffer contains the entries "A" and "C" and the `Marker` is "B", the `Marker` will **not** be found and the `IndexNum` will be 0 on return from `FspFileSystemSearchDirectoryBuffer`. It will then be incremented to 1 by the code at this [location](https://github.com/billziss-gh/winfsp/blob/v1.8/src/dll/dirbuf.c#L378), which also seems correct.

**Comment by pmaxwell-iam** [comment]
The issue being referred to is where the buffer contains the entries "B" and "C" and the Marker is "A". In this case the IndexNum will be returned as 0, incremented to 1, and "B" will be skipped.

**Comment by billziss-gh** [comment]
>The context was a folder deletion where the earlier files in the folder have already been removed. The test case had 100 files in the folder; having deleted 0-53 it then called QueryDirectory with marker "53.txt", got back a buffer containing 54-99 and continued by deleting 55 onward.

BTW, if `QueryDirectory` returned 54.txt-99.txt after getting the marker 53.txt, this is actually correct. `QueryDirectory` is supposed to return all entries **after** the marker.

Here is an alternative scenario that may or may not be correct:

- QueryDirectory returned correct results.
- The program driving the deletions received correct results and sent `DeleteFile` or equivalent requests to the file system.
- The passthrough file system attempted to delete the file `54.txt`. However passthrough uses `HANDLE`'s to manage files and Windows restricts the operations that can be performed on file `HANDLE`'s. For example, the OS and file system drivers may not delete files immediately when the file disposition flag is set and the file is already opened by another app. This could mean that the passthrough file system was unable to delete the file `54.txt` in this case.

To make this scenario more clear, consider what happens when a file is opened in passthrough. In the following OP is the "Originating Process" (the process that tries to open), FS is the passthrough file system and NTFS is the actual disk file system where the passthrough files are stored.

- OP issues `CreateFile("X:\foo\54.txt")`.
- FS receives `Open("\foo\54.txt")`.
- FS issues `CreateFile("C:\PREFIX\foo\54.txt")`. This `CreateFile` call is subject to normal Windows restrictions  (e.g. access control, delete pending, etc.) So it may fail for reasons that are expected for a regular Windows process, but **not** expected for a file system.
- NTFS receives `IRP_MJ_CREATE` and responds with SUCCESS. Windows OS converts this to a `HANDLE` and returns.
- FS receives `HANDLE` -- call it H<sub>fs</sub>.
- FS returns success. Windows OS converts this to a `HANDLE` and returns.
- OP receives `HANDLE` -- call it H<sub>op</sub>.

Now every operation on H<sub>op</sub> will be reflected on an equivalent operation on H<sub>fs</sub>. There are two problems with this approach:

- Because the file `"C:\PREFIX\foo\54.txt"` is accessible to other Windows software directly, it is possible that it may be opened by someone outside the passthrough file system. This may restrict the kind of operations that can be performed on H<sub>fs</sub>.

- Windows will often send operations to file system drivers that are not strictly "legal" from an application point of view. For example, I believe it is possible for Windows to send a `Write` request to a file that was opened for read-only access. (I believe this can happen when using the NTOS Cache Manager: an application opens a file for read access and receives handle H<sub>a</sub>; the Cache Manager registers the HANDLE internally for its own purposes; the application opens the same file for read-write access and receives handle H<sub>b</sub>; it proceeds to write to H<sub>b</sub> and data is buffered inside the Cache Manager; at a later time the Cache Manager flushes the data via H<sub>a</sub>; this would of course fail in passthough.)

For this reason in general I do not recommend use of passthrough in production. It was written as an example file system for a tutorial and has many similar problems.

**Comment by pmaxwell-iam** [comment]
Apologies, I may not have been sufficiently clear. I was using BufferedReadDirectory from the unsupported winfsp.hpp.

This fills the directory buffer using ReadDirectoryEntry, and then calls FspFileSystemReadDirectoryBuffer. The buffer it was reading contained entries 54-99, but since the marker was set to "53.txt", it skipped 54 and went to 55.

**Comment by pmaxwell-iam** [comment]
The passthrough filesystem avoids this problem by keeping a per-handle directory buffer which reflects the historic rather than actual state of the directory - it retains the list of children gathered on the first query even if the files under the directory are then deleted. This list can safely be queried because the marker file will still be present in this cache by the time of the next ReadDirectory call with a marker, even if it is no longer present on the filesystem.

There are obviously some questions as to what would happen in this instance if the same handle was used to query from the beginning again, or if a different handle were used to call `ReadDirectory` with a marker; I don't know if either of these scenarios are possible.

However, the issue seems clearly to be with `FspFileSystemReadDirectoryBuffer` rather than `BufferedReadDirectory` in this instance.

**Comment by billziss-gh** [comment]
> The issue being referred to is where the buffer contains the entries "B" and "C" and the Marker is "A". In this case the IndexNum will be returned as 0, incremented to 1, and "B" will be skipped.

Hmmm... I would not expect this to be possible because the FSD will only send back markers that it has seen in prior `QueryDirectory` results in the same `FileDesc` (i.e. for the same HANDLE).

As you say:

> The passthrough filesystem avoids this problem by keeping a per-handle directory buffer which reflects the historic rather than actual state of the directory

So the FSD should not send a marker not already in the directory buffer (when using directory buffers).

----

But in any case if this is indeed possible, it needs to be fixed. One possible fix would be to check the return value of `FspFileSystemSearchDirectoryBuffer` and if it is `FALSE` and the value of `IndexNum` is also 0, then `IndexNum` should not be incremented. But I cannot help but feel that this may or may not be a complete fix for the problem and that it may introduce additional problems elsewhere.

There is a test in [dirbuf-test.c](https://github.com/billziss-gh/winfsp/blob/master/tst/winfsp-tests/dirbuf-test.c) which should be augmented as well.

**Comment by pmaxwell-iam** [comment]
>I would not expect this to be possible because the FSD will only send back markers that it has seen in prior QueryDirectory results in the same FileDesc (i.e. for the same HANDLE).

The filesystem may have seen the file previously, but that does not mean that it still exists.

In the case that a recursive directory deletion is being carried out it is almost certain that the marker file has been deleted at the point the QueryDirectory call with marker is issued. If the buffer reflects the current state rather than the state when the first QueryDirectory call was made on that handle, the marker file will not exist in the buffer.

The binary search is incapable of returning the first entry even if the thing being searched for comes before the first file. I believe we're fine in the other direction (if the thing being searched for is beyond the last file) - it seems to me IndexNum should exceed Count and we add no DirInfos - but I haven't tested this. Assuming this is correct I think your proposed fix should resolve the issue completely.

Happy to make the change and submit a PR, but if you'd prefer to take a bit more time to study it yourself before committing to a course of action I'll hold off until you're satisfied.

**Comment by billziss-gh** [comment]
> In the case that a recursive directory deletion is being carried out it is almost certain that the marker file has been deleted at the point the QueryDirectory call with marker is issued. If the buffer reflects the current state rather than the state when the first QueryDirectory call was made on that handle, the marker file will not exist in the buffer.

Understood.

However I note that the directory buffer is not intended to be used like this. Instead the directory buffer is intended to always reflect the directory state when the first `ReadDirectory` for a particular descriptor arrives. If this is observed the FSD will never send a marker that does not exist inside the directory buffer (although the file named by the marker may no longer exist in the actual file system -- but it is ok to return stale directory contents).

In more detail the directory buffer is intended to be used as follows:

- Upon receiving an `Open` request for a directory, the file system sets up an empty (null) void pointer `DirBuffer` that can be used to hold a directory buffer (if one is needed). It returns this as part of the "file descriptor" (see the description [here](https://github.com/billziss-gh/winfsp/blob/v1.8/inc/winfsp/winfsp.h#L177-L182) -- this means that different `Open` requests will receive different file descriptors and thus different directory buffers).

- Upon receiving the first `ReadDirectory` request for a particular file descriptor or whenever the marker is 0, the file system fills up the directory buffer. Subsequent `ReadDirectory` requests will be satisfied from the filled buffer. (This can of course mean that `ReadDirectory` may return an entry for a file that no longer exists -- IMO this is not a problem.) The general pattern used to fill and use the directory buffer is as follows (from passthrough):

    ```C
    if (FspFileSystemAcquireDirectoryBuffer(&FileContext->DirBuffer, 0 == Marker, &DirBufferResult))
    {
        loop_and_fill_the_directory_buffer:
            FspFileSystemFillDirectoryBuffer(&FileContext->DirBuffer, DirInfo, &DirBufferResult);

        FspFileSystemReleaseDirectoryBuffer(&FileContext->DirBuffer);
    }

    FspFileSystemReadDirectoryBuffer(&FileContext->DirBuffer,
        Marker, Buffer, BufferLength, PBytesTransferred);
    ```

  - Note that if you refill the directory buffer in cases where the marker is non-zero, then it is of course possible to receive a non-existent marker.

- Upon receiving a `Close` request the file system deletes any directory buffer using `FspFileSystemDeleteDirectoryBuffer`.

> The binary search is incapable of returning the first entry even if the thing being searched for comes before the first file.

I agree.

Again I do not believe this to be a problem in the way that the directory buffer is supposed to be used. However if you are seeing this perhaps I am overlooking a case where this can happen. This is why it would be great if we had a repro.

> I believe we're fine in the other direction (if the thing being searched for is beyond the last file) - it seems to me IndexNum should exceed Count and we add no DirInfos - but I haven't tested this.

I agree.

> Happy to make the change and submit a PR, but if you'd prefer to take a bit more time to study it yourself before committing to a course of action I'll hold off until you're satisfied.

I think my biggest issue with this is that I still cannot see how a non-existent marker can be sent. If you are seeing this in regular passthrough without any other changes, I cannot help but wonder if the actual problem is something else (see my earlier discussion about `HANDLE` usage in passthrough).

In any case I propose the following:

- If possible, consider setting a breakpoint at this [`FspFileSystemSearchDirectoryBuffer` location](https://github.com/billziss-gh/winfsp/blob/v1.8/src/dll/dirbuf.c#L117). If I am right, this breakpoint should never hit with unmodified passthrough.

- If possible, consider adding a debug log message at the same [`FspFileSystemSearchDirectoryBuffer` location](https://github.com/billziss-gh/winfsp/blob/v1.8/src/dll/dirbuf.c#L117). For example, `FspDebugLog(__FUNCTION__ "\n")` should do it. This will appear either in DbgView or in your console window if you start (unmodified) passthrough with the `-D -` option.

Finally, and especially, if you find that the breakpoint or debug log message gets hit, you can try the **untested** patch below:

```patch
diff --git a/src/dll/dirbuf.c b/src/dll/dirbuf.c
index d2605578..f4064b23 100644
--- a/src/dll/dirbuf.c
+++ b/src/dll/dirbuf.c
@@ -115,6 +115,15 @@ static BOOLEAN FspFileSystemSearchDirectoryBuffer(FSP_FILE_SYSTEM_DIRECTORY_BUFF
     }

     *PIndexNum = Lo;
+    if (0 == Lo && (ULONG)Lo < Count)
+    {
+        DirInfo = (PVOID)(DirBuffer->Buffer + Index[Lo]);
+        CmpResult = FspFileSystemDirectoryBufferFileNameCmp(
+            DirInfo->FileNameBuf, (DirInfo->Size - sizeof *DirInfo) / sizeof(WCHAR),
+            Marker, MarkerLen);
+        if (0 < CmpResult)
+            *PIndexNum--;
+    }
     return FALSE;
 }
```

This patch should correctly handle both cases when `IndexNum == 0`:

- If the buffer contains A, C and we search for B, the correct `IndexNum` is `0`.
- If the buffer contains B, C and we search for A, the correct `IndexNum` is `-1` (or `0xffffffff`, since `IndexNum` is a `ULONG`).

**Comment by pmaxwell-iam** [comment]
Thanks for the lengthy reply. This helps clarify things for me.

My initial reproduction was in a filesystem I was attaching to the API using the same interface as passthrough - likely this will get moved to make use of the C interface later, but the C++ interface seemed a simpler beginning before I realised it was not fully supported. In this filesystem I was returning the current contents of the directory in the directory buffer - hence the problem.

I believed I had replicated this on passthrough, but carelessly did not keep the logs (I very much appreciate the detail level of your logs, but they grow in size very quickly). Subsequent attempts did not replicate. My primary hypothesis is that I was mistaken.

> This can of course mean that ReadDirectory may return an entry for a file that no longer exists -- IMO this is not a problem

I was initially somewhat concerned about this, though from the perspective of concurrency it's perfectly possible for another process to delete or add files after you perform your query. Mostly I was not sure what guarantees existed that ReadDirectory would not be called with a marker on a handle that had not previously called ReadDirectory without a marker, or that the same handle would not be reused for more than one query as this would make the staleness more problematic.

For the time being I have adopted a similar approach with a persistent buffer attached to the handle.

**Comment by billziss-gh** [comment]
> Mostly I was not sure what guarantees existed that ReadDirectory would not be called with a marker on a handle that had not previously called ReadDirectory without a marker

(In the following I use "file node" and "file descriptor" terminology from [here](https://github.com/billziss-gh/winfsp/blob/v1.8/inc/winfsp/winfsp.h#L165-L189).)

When you track files using file descriptors you will always get a `ReadDirectory` with `Marker == 0` first for a particular file descriptor. You can optionally use directory buffers to implement `ReadDirectory` functionality, as explained in my previous message.

When you track files using file nodes there is no guarantee about the `ReadDirectory` markers that you receive. The expectation is that the file system will go into its own directory data structures and use the marker to determine which directory entries to return. You cannot use directory buffers in this case. For an example of such a file system, see MEMFS.

>  same handle would not be reused for more than one query as this would make the staleness more problematic

The Win32 API for fetching directory entries is `FindFirstFile` and friends. In general applications will use it in a loop to enumerate files and then close the directory handle using `FindClose`, but it is of course possible (although not very productive) to keep the handle open indefinitely.

I personally am not worried about such cases. As you say: "though from the perspective of concurrency it's perfectly possible for another process to delete or add files after you perform your query".

> For the time being I have adopted a similar approach with a persistent buffer attached to the handle.

This should work. If you find otherwise, we can follow up.

**Comment by billziss-gh** [comment]
@pmaxwell-iam I am planning to close this as it seems resolved. Please let me know if you think otherwise.

**Comment by pmaxwell-iam** [comment]
I think at the least it would be good to document the intended usage of the buffer code, as the binary search currently has a bug which is only avoided if you retain the initial buffer. Not only is this not documented anywhere, the example code in `winfsp.hpp` has a ReadDirectoryEntry method which supplies a marker as a parameter; if this marker parameter is used to filter within ReadDirectoryEntry, the bug manifests.

I am not currently hindered now I have a workaround, but given it took me some time to discover the issue and correspondence to confirm how things were supposed to work, I feel either documenting the function usage or fixing the binary search bug would be useful for other developers against the API in future.

**Comment by billziss-gh** [comment]
@pmaxwell-iam 

It turns out that I was wrong and the described problem can manifest itself under the following conditions:

- The FSD has cached a portion of the directory listing.
- The file system supports directory pattern matching (`PassQueryDirectoryPattern==TRUE`).

I was able to reproduce this with passthrough, which can do directory pattern matching. I have added a fix and a new test case (`querydir_single_test`) for this issue.
